File: frontend/src/app/api/haversine.ts
const EARTH_RADIUS = 6371e3;


function convertDegreesToRadians(degrees: number): number
{
    return degrees * Math.PI / 180;
}


export default function haversine(
    lat1: number, lon1: number, lat2: number, lon2:number
): number
{
    const deltaLat = convertDegreesToRadians(lat2 - lat1);
    const deltaLon = convertDegreesToRadians(lon2 - lon1);
    const a = 
        Math.sin(deltaLat / 2) ** 2 +
        Math.cos(convertDegreesToRadians(lat1)) *
        Math.cos(convertDegreesToRadians(lat2)) *
        Math.sin(deltaLon / 2) ** 2;
    const c = 2 * Math.atan2(
        Math.sqrt(a),
        Math.sqrt(1 - a)
    );
    return EARTH_RADIUS * c;
}

File: packages/py/zipcoords/haversine.py
import math

def haversine(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    R = 6371e3
    to_rad = math.radians

    φ1 = to_rad(lat1)
    φ2 = to_rad(lat2)
    Δφ = to_rad(lat2 - lat1)
    Δλ = to_rad(lon2 - lon1)

    a = \
        math.sin(Δφ / 2) ** 2 + \
        math.cos(φ1) * \
        math.cos(φ2) * \
        math.sin(Δλ / 2) ** 2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

    return R * c
